src/nl/hannahsten/texifyidea/refactoring/introduceCommand/LatexExtractCommandHandler.kt

```diff
@@ -0,0 +1,370 @@
+    expr: LatexExtractablePSI,
+    occurrences: List<LatexExtractablePSI>,
+    callback: (List<LatexExtractablePSI>) -> Unit
+) {
+    if (isUnitTestMode && occurrences.size > 1) {
+        callback(MOCK!!.chooseOccurrences(expr, occurrences))
+    }
+    else {
+        OccurrencesChooser.simpleChooser<PsiElement>(editor)
+            .showChooser(
+                expr,
+                occurrences,
+                { choice: OccurrencesChooser.ReplaceChoice ->
+                    val toReplace = if (choice == OccurrencesChooser.ReplaceChoice.ALL) occurrences else listOf(expr)
+                    callback(toReplace)
+                }.asPass
+            )
+    }
+}
+
+private val <T> ((T) -> Unit).asPass: Pass<T>
+    get() = object : Pass<T>() {
+        override fun pass(t: T) = this@asPass(t)
+    }
+
+fun findExpressionInRange(file: PsiFile, startOffset: Int, endOffset: Int): LatexExtractablePSI? {
+    val firstUnresolved = file.findElementAt(startOffset) ?: return null
+    val first =
+        if (firstUnresolved is PsiWhiteSpace)
+            file.findElementAt(firstUnresolved.endOffset) ?: return null
+        else
+            firstUnresolved
+
+    val lastUnresolved = file.findElementAt(endOffset - 1) ?: return null
+    val last =
+        if (lastUnresolved is PsiWhiteSpace)
+            file.findElementAt(lastUnresolved.startOffset - 1) ?: return null
+        else
+            lastUnresolved
+
+    val parent = findCommonParent(first, last) ?: return null
+
+    return if (parent is LatexNormalText) {
+        LatexExtractablePSI(parent, TextRange(startOffset - parent.startOffset, endOffset - parent.startOffset))
+    }
+    else
+        LatexExtractablePSI(parent)
+}
+
+fun findCandidateExpressionsToExtract(editor: Editor, file: LatexFile): List<LatexExtractablePSI> {
+    val selection = editor.selectionModel
+    if (selection.hasSelection()) {
+        // If there's an explicit selection, suggest only one expression
+        return listOfNotNull(findExpressionInRange(file, selection.selectionStart, selection.selectionEnd))
+    }
+    else {
+        val expr = findExpressionAtCaret(file, editor.caretModel.offset)
+            ?: return emptyList()
+        if (expr is LatexBeginCommand) {
+            val endCommand = expr.endCommand()
+            if (endCommand == null)
+                return emptyList()
+            else {
+                val environToken = findCommonParent(expr, endCommand)
+                return if (environToken != null)
+                    listOf(LatexExtractablePSI(environToken))
+                else
+                    emptyList()
+            }
+        }
+        else if (expr is LatexNormalText) {
+            return listOf(LatexExtractablePSI(expr))
+        }
+        else {
+            if (expr.elementType == NORMAL_TEXT_WORD) {
+                val interruptedParent = expr.firstParentOfType(LatexNormalText::class)
+                    ?: throw IllegalStateException("You suck")
```

Comment:
I'm not sure why you specifically need a LatexNormalText, but NORMAL_TEXT_WORD can appear in other psi elements, like parameter text.

Reply:
Can you give an example where it breaks? 

The point is really to just look harder for other offerings to extract. It may not find anything else, but it can look.

Reply:
Sure: `\text{4<caret>2}`

Reply:
I am going to leave the exception for now so it is way more obvious when it fails. I will resolve this when we are more confident that this is the way to go.